Remove nextToken from ModelPagination since nextToken is never exposed

richardmcclellan opened the pull request:
`nextToken` is an AppSync detail, which should not be exposed in the public interface.   There is no way to obtain it from a response directly, so we can remove it from the `ModelPagination` object.

To make a request to get a "next page" of results, customers should obtain a `GraphQLRequest` via the `getRequestForNextPage()` method on the `PaginatedResult` object.  This request object is a copy of the `GraphQLRequest` used in the previous query, but with the `nextToken` added as a variable.   Then, customers can pass this GraphQLRequest directly to `Amplify.API.query(...)`

This also adds an additional factory method, so customers can do `ModelPagination.limit(1000)`, instead of `ModelPagination.firstPage().withLimit(1000)`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

raphkim commented:
> To make a request to get a "next page" of results, customers should obtain a `GraphQLRequest` via the `getRequestForNextPage()` method on the `PaginatedResult` object. This request object is a copy of the `GraphQLRequest` used in the previous query, but with the `nextToken` added as a variable. Then, customers can pass this GraphQLRequest directly to `Amplify.API.query(...)`

Is this behavior tested anywhere? 

Also, one random thing I noticed while looking through the code: the constructor for `AppSyncPaginatedResult` seems to take in `@NonNull GraphQLRequest<PaginatedResult<T>> requestForNextResult` but its implementation of `hasNextResult()` is `return requestForNextResult != null;`, which always evaluates to `true`. Will `requestForNextResult` always be non-null as the annotation suggests?

richardmcclellan commented:
> > To make a request to get a "next page" of results, customers should obtain a `GraphQLRequest` via the `getRequestForNextPage()` method on the `PaginatedResult` object. This request object is a copy of the `GraphQLRequest` used in the previous query, but with the `nextToken` added as a variable. Then, customers can pass this GraphQLRequest directly to `Amplify.API.query(...)`
> 
> Is this behavior tested anywhere?
> 

Good callout, turns out it's not.   I just added a test, and it actually caught a bug where an exception was thrown whenever a paginated query was made.  I fixed the bug as well.  Yay tests!

> Also, one random thing I noticed while looking through the code: the constructor for `AppSyncPaginatedResult` seems to take in `@NonNull GraphQLRequest<PaginatedResult<T>> requestForNextResult` but its implementation of `hasNextResult()` is `return requestForNextResult != null;`, which always evaluates to `true`. Will `requestForNextResult` always be non-null as the annotation suggests?

Good call, it can be null, so I changed it to `@Nullable`

jamesonwilliams commented on aws-api/src/test/java/com/amplifyframework/api/aws/AWSApiPluginTest.java:
```diff
@@ -158,6 +161,32 @@ public void graphQlQueryRendersValidResponse() throws ApiException {
     }
 
+    /**
+     * Same as graphQlQueryRendersValidResponse, except with pagination.   Expect a PaginatedResult&lt;BlogOwner&gt;
```
```suggestion
     * Same as {@link #graphQlQueryRendersValidResponse()}, except with pagination.
     * Expect a PaginatedResult&lt;BlogOwner&gt;
```

jamesonwilliams commented on aws-api/src/main/java/com/amplifyframework/api/graphql/model/ModelQuery.java:
```diff
@@ -122,7 +122,7 @@ private ModelQuery() {}
     ) {
         return AppSyncGraphQLRequestFactory.buildPaginatedResultQuery(
-                modelType, null, pagination.getLimit());
+                modelType, QueryPredicates.all(), pagination.getLimit());
```
Woop!

jamesonwilliams commented on aws-api/src/main/java/com/amplifyframework/api/aws/AppSyncPaginatedResult.java:
```diff
@@ -27,7 +28,7 @@
 
     AppSyncPaginatedResult(@NonNull Iterable<T> items,
-                           @NonNull GraphQLRequest<PaginatedResult<T>> requestForNextResult) {
+                           @Nullable GraphQLRequest<PaginatedResult<T>> requestForNextResult) {
```
Is it possible to have a paginated result, that has no request associated with it?

richardmcclellan replied:
Yes, if `nextToken` is null (the last page), then the `requestForNextResult` will also be null
